compiler/rustc_mir/src/interpret/cast.rs

```diff
@@ -269,12 +269,61 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                     Immediate::new_slice(ptr, length.eval_usize(*self.tcx, self.param_env), self);
                 self.write_immediate(val, dest)
             }
-            (&ty::Dynamic(..), &ty::Dynamic(..)) => {
-                // For now, upcasts are limited to changes in marker
-                // traits, and hence never actually require an actual
-                // change to the vtable.
+            (&ty::Dynamic(ref data_a, ..), &ty::Dynamic(ref data_b, ..)) => {
                 let val = self.read_immediate(src)?;
-                self.write_immediate(*val, dest)
+                if data_a.principal_def_id() == data_b.principal_def_id() {
+                    self.write_immediate(*val, dest)
+                } else {
+                    // trait upcasting coercion
+                    let principal_a = data_a.principal().expect(
+                        "unsize_into_ptr: missing principal trait for trait upcasting coercion",
+                    );
+                    let principal_b = data_b.principal().expect(
+                        "unsize_into_ptr: missing principal trait for trait upcasting coercion",
+                    );
+
+                    let vptr_entry_idx = self.tcx.vtable_trait_upcasting_coercion_new_vptr_slot((
+                        principal_a.with_self_ty(*self.tcx, src_pointee_ty),
+                        principal_b.def_id(),
+                    ));
+
+                    if let Some(entry_idx) = vptr_entry_idx {
+                        let (old_data, old_vptr) = match *val {
+                            Immediate::ScalarPair(data, vptr) => (data, vptr),
+                            Immediate::Scalar(..) => span_bug!(
+                                self.cur_span(),
+                                "{:?} input to a trait upcasting cast ({:?} -> {:?})",
+                                *src,
+                                source_ty,
+                                cast_ty
+                            ),
+                        };
+                        let pointer_size = self.pointer_size();
+                        let entry_idx = u64::try_from(entry_idx).unwrap();
+                        let old_data = old_data.check_init()?;
+                        let old_vptr = self.scalar_to_ptr(old_vptr.check_init()?);
+                        let vtable = self
```

Comment:
Please don't interact with the vtable directly in this file. Instead, add a new method in `traits.rs` (similar to how we have `get_vtable_slot` to read the fn ptr at a particular slot from the vtable).

Reply:
Done.